gateway/src/apicast/policy/headers.lua

```diff
@@ -0,0 +1,129 @@
+--- Headers policy
+-- This policy allows to include custom headers that will be sent to the
+-- upstream as well as modify or delete the ones included in the original
+-- request.
+-- Similarly, this policy also allows to add, modify, and delete the headers
+-- included in the response.
+
+local ipairs = ipairs
+local type = type
+local insert = table.insert
+
+local policy = require('apicast.policy')
+local _M = policy.new('Headers policy')
+
+local new = _M.new
+
+local function new_header_value(current_value, value_to_add)
+  local new_value = current_value or {}
+
+  if type(new_value) == 'string' then
+    new_value = { new_value }
+  end
+
+  insert(new_value, value_to_add)
+  return new_value
+end
+
+local function add_request_header(header, value, req_headers)
+  local new_value = new_header_value(req_headers[header], value)
+  ngx.req.set_header(header, new_value)
+end
+
+local function set_request_header(header, value)
+  ngx.req.set_header(header, value)
+end
+
+local function push_request_header(header, value, req_headers)
+  if req_headers[header] then
+    add_request_header(header, value, req_headers)
+  end
+end
+
+local function add_resp_header(header, value)
+  local new_value = new_header_value(ngx.header[header], value)
```

Comment:
I'd consider renaming `header` to `name` or `header_name`. Because it looks like there is too much use of `header` (because of the `ngx.header`).

Reply:
Good idea :+1: